File: js/039_CombinationSum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    var ret = [];
    candidates.sort(function(a,b){return a - b;});
    backtrack(ret,[],candidates,target,0);
    return ret;
};

var backtrack = function(ret, path, candidates, target,start){
    if(target === 0){
        ret.push(path.concat());
        return;
    }
    if(target < 0) return;
    for(var i = start; i < candidates.length; ++i){
        path.push(candidates[i]);
        backtrack(ret,path,candidates,target-candidates[i], i);
        path.pop();
    }
}

File: cpp/039_CombinationSum.cpp
 class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ret;
        vector<int> path;
        std::sort(candidates.begin(),candidates.end());
        backtrack(ret,path,candidates,target,0);
        return ret;
    }
private:
    void backtrack(vector<vector<int>>& ret, vector<int>& path,vector<int>& candidates, int target,int start){
        if(target == 0){         
            ret.push_back(path);
            return;
        }
        if(target < 0){
            return;
        }
        for(int i = start; i < candidates.size(); ++i){
            path.push_back(candidates[i]);
            backtrack(ret,path,candidates,target-candidates[i],i);
            path.pop_back();
        }
    }
};